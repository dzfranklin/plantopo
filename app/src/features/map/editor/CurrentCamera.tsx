import type * as ml from 'maplibre-gl';

export interface CameraBBox {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
}

export class CurrentCameraPosition {
  private constructor(
    public center: [number, number],
    public zoom: number,
    public bearing: number,
    public pitch: number,
    public bbox: CameraBBox,
    private _map: ml.Map, // Hack: So we can use for projection
  ) {}

  static fromMap(map: ml.Map): CurrentCameraPosition {
    const center = map.getCenter();
    const bounds = map.getBounds();
    return new CurrentCameraPosition(
      [center.lng, center.lat],
      map.getZoom(),
      map.getBearing(),
      map.getPitch(),
      {
        minY: bounds.getNorth(),
        maxY: bounds.getSouth(),
        minX: bounds.getEast(),
        maxX: bounds.getWest(),
      },
      map,
    );
  }

  // Hack: I should use https://github.com/proj4js/proj4js instead of keeping a
  // ref to map so it doesn't have to be just the current camera we can project
  // in. But remember we need to account for pitch and bearing

  project(lnglat: [number, number] | GeoJSON.Position): [number, number] {
    if (lnglat.length < 2) {
      // GeoJSON.Position must be length 2 or 3
      throw new Error('not a lnglat');
    }
    const pt = this._map.project([lnglat[0], lnglat[1]]);
    return [pt.x, pt.y];
  }

  unproject(xy: [number, number]): [number, number] {
    const lnglat = this._map.unproject(xy);
    return [lnglat.lng, lnglat.lat];
  }
}
