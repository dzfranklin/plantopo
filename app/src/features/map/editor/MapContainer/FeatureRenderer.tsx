import { CameraPosition } from '../CameraPosition';
import { Scene } from '../api/SyncEngine/Scene';

interface RenderTree {
  camera: CameraPosition;
}

export class Renderer {
  render(scene: Scene, camera: CameraPosition): RenderTree {
    return { camera };
  }
}
