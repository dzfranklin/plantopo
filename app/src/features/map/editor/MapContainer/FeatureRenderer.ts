import { BBoxPolygon } from '../CurrentCamera';
import { Scene, SceneFeature, SceneRootFeature } from '../engine/Scene';
import { booleanContains } from '@turf/turf';
import booleanIntersects from '@turf/boolean-intersects';
import { SyncGeometry } from '@/gen/sync_schema';
import { midpoint2 } from '@/generic/vector2';

export interface RenderList {
  timing: {
    scene: Scene['timing'];
    start: number;
    end: number;
  };
  list: RenderItem[];
}

export type RenderItem = RenderFeature | RenderFeatureHandle;

export interface RenderFeature {
  type: 'feature';
  id: string;
  children: RenderFeature[];

  active: boolean;
  selectedByMe: boolean;
  selectedByPeers: string[] | null;
  hoveredByMe: boolean;

  geometry: SyncGeometry;

  name: string | null;
  color: string;
}

export interface RenderFeatureHandle {
  type: 'handle';
  id: string; // id of the handle (not the id of the feature)
  feature: RenderFeature;
  handleType: 'vertex' | 'midpoint';
  geometry: {
    type: 'Point';
    coordinates: [number, number];
  };
}

interface Inherited {
  color: string;
}

const ROOT_INHERITED: Inherited = { color: '#000000' };

export class FeatureRenderer {
  render(scene: Scene, clipBox: BBoxPolygon): RenderList {
    const start = performance.now();

    const list: RenderFeature[] = [];
    this._render(clipBox, ROOT_INHERITED, scene.features.root, list);
    const end = performance.now();
    return { list, timing: { scene: scene.timing, start, end } };
  }

  private _render(
    clipBox: BBoxPolygon,
    inherited: Inherited,
    feature: SceneRootFeature | SceneFeature,
    out: RenderItem[],
  ) {
    const [newInherited, itself] = this._renderItself(
      clipBox,
      inherited,
      feature,
    );
    for (const item of itself) {
      out.push(item);
    }

    for (const child of feature.children) {
      this._render(clipBox, newInherited ?? inherited, child, out);
    }
  }

  private _renderItself(
    clipBox: BBoxPolygon,
    inherited: Inherited,
    feature: SceneRootFeature | SceneFeature,
  ): [Inherited | null, RenderItem[]] {
    if (!feature.parent) {
      // The root isn't rendered itself
      return [null, []];
    }

    let newInherited = null;
    if (feature.color) {
      newInherited = { ...inherited, color: feature.color };
    }

    if (!feature.geometry || feature.hidden) {
      return [newInherited, []];
    }

    let geometry: SyncGeometry;
    if (feature.geometry.type === 'Point') {
      if (!booleanContains(clipBox, feature.geometry)) {
        return [newInherited, []];
      }
      geometry = feature.geometry;
    } else if (feature.geometry.type === 'LineString') {
      if (!booleanIntersects(feature.geometry, clipBox)) {
        return [newInherited, []];
      }
      geometry = feature.geometry;
    } else {
      return [newInherited, []];
    }

    const itself: RenderItem[] = [];

    const rf: RenderFeature = {
      ...feature,
      type: 'feature',
      children: [],
      geometry,
      color: feature.color ?? inherited.color,
    };
    itself.push(rf);

    if (feature.geometry.type === 'LineString') {
      const coords = geometry.coordinates as [number, number][];
      for (const [i, coord] of coords.entries()) {
        itself.push({
          type: 'handle',
          feature: rf,
          id: `${feature.id}-handle${i}`,
          handleType: 'vertex',
          geometry: {
            type: 'Point',
            coordinates: coord,
          },
        });

        if (i < coords.length - 1) {
          const nextCoord = coords[i + 1]!;
          itself.push({
            type: 'handle',
            feature: rf,
            id: `${feature.id}-handle${i}-midpoint`,
            handleType: 'midpoint',
            geometry: {
              type: 'Point',
              coordinates: midpoint2(coord, nextCoord),
            },
          });
        }
      }
    }

    return [newInherited, itself];
  }
}
