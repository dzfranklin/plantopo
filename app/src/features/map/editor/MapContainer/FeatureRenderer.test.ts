import { bboxPolygonOf } from '../CurrentCamera';
import {
  EMPTY_SCENE,
  Scene,
  SceneFeature,
  SceneRootFeature,
} from '../engine/Scene';
import { FeatureRenderer } from './FeatureRenderer';

let subject: FeatureRenderer;
let idOffset: number;
beforeEach(() => {
  idOffset = 0;
  subject = new FeatureRenderer();
});

test('single point snapshot', () => {
  const got = subject.render(
    sceneFixture({
      features: {
        children: [
          {
            geometry: {
              type: 'Point',
              coordinates: [50, 50],
            },
          },
        ],
      },
    }),
    bboxPolygonOf(0, 0, 100, 100),
  );
  expect(got.list).toMatchSnapshot();
});

test('single line snapshot', () => {
  const got = subject.render(
    sceneFixture({
      features: {
        children: [
          {
            geometry: {
              type: 'LineString',
              coordinates: [
                [50, 50],
                [60, 60],
              ],
            },
          },
        ],
      },
    }),
    bboxPolygonOf(0, 0, 100, 100),
  );
  expect(got.list).toMatchSnapshot();
});

test('line render has handles', () => {
  const got = subject.render(
    sceneFixture({
      features: {
        children: [
          {
            geometry: {
              type: 'LineString',
              coordinates: [
                [50, 50],
                [60, 60],
              ],
            },
          },
        ],
      },
    }),
    bboxPolygonOf(0, 0, 100, 100),
  ).list;
  expect(
    got.filter((f) => f.type === 'handle' && f.handleType === 'vertex'),
  ).toHaveLength(2);
  expect(
    got.filter((f) => f.type === 'handle' && f.handleType === 'midpoint'),
  ).toHaveLength(1);
});

function idFixture(): string {
  return `id${++idOffset}`;
}

interface SceneTemplate {
  features: RootFeatureTemplate;
}

type RootFeatureTemplate = { children: Array<FeatureTemplate> };

type FeatureTemplate = Partial<Omit<SceneFeature, 'parent'>>;

function sceneFixture(tmpl: SceneTemplate): Scene {
  const rootFeature: SceneRootFeature = { id: '', parent: null, children: [] };
  return {
    ...EMPTY_SCENE,
    features: {
      ...EMPTY_SCENE.features,
      root: {
        id: '',
        parent: null,
        children: tmpl.features.children.map((tmpl) =>
          sceneFixtureFeature(tmpl, rootFeature),
        ),
      },
    },
  };
}

function sceneFixtureFeature(
  tmpl: FeatureTemplate,
  parent: SceneRootFeature | SceneFeature,
): SceneFeature {
  const f = {
    parent,
    idx: '',
    children: [],
    hidden: false,

    active: false,
    selectedByMe: false,
    selectedByPeers: null,
    hoveredByMe: false,

    geometry: null,

    name: null,
    color: null,

    ...tmpl,

    id: tmpl.id ?? idFixture(),
  };
  if (tmpl.children) {
    f.children = tmpl.children.map((tmpl) => sceneFixtureFeature(tmpl, f));
  }
  return f;
}
