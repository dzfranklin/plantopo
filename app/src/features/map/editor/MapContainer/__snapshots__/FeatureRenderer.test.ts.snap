// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single line snapshot 1`] = `
[
  {
    "active": false,
    "children": [],
    "color": "#000000",
    "geometry": {
      "coordinates": [
        [
          50,
          50,
        ],
        [
          60,
          60,
        ],
      ],
      "type": "LineString",
    },
    "hidden": false,
    "hoveredByMe": false,
    "id": "id1",
    "idx": "",
    "name": null,
    "parent": {
      "children": [],
      "id": "",
      "parent": null,
    },
    "selectedByMe": false,
    "selectedByPeers": null,
    "type": "feature",
  },
  {
    "feature": {
      "active": false,
      "children": [],
      "color": "#000000",
      "geometry": {
        "coordinates": [
          [
            50,
            50,
          ],
          [
            60,
            60,
          ],
        ],
        "type": "LineString",
      },
      "hidden": false,
      "hoveredByMe": false,
      "id": "id1",
      "idx": "",
      "name": null,
      "parent": {
        "children": [],
        "id": "",
        "parent": null,
      },
      "selectedByMe": false,
      "selectedByPeers": null,
      "type": "feature",
    },
    "geometry": {
      "coordinates": [
        50,
        50,
      ],
      "type": "Point",
    },
    "handleType": "vertex",
    "id": "id1-handle0",
    "type": "handle",
  },
  {
    "feature": {
      "active": false,
      "children": [],
      "color": "#000000",
      "geometry": {
        "coordinates": [
          [
            50,
            50,
          ],
          [
            60,
            60,
          ],
        ],
        "type": "LineString",
      },
      "hidden": false,
      "hoveredByMe": false,
      "id": "id1",
      "idx": "",
      "name": null,
      "parent": {
        "children": [],
        "id": "",
        "parent": null,
      },
      "selectedByMe": false,
      "selectedByPeers": null,
      "type": "feature",
    },
    "geometry": {
      "coordinates": [
        55,
        55,
      ],
      "type": "Point",
    },
    "handleType": "midpoint",
    "id": "id1-handle0-midpoint",
    "type": "handle",
  },
  {
    "feature": {
      "active": false,
      "children": [],
      "color": "#000000",
      "geometry": {
        "coordinates": [
          [
            50,
            50,
          ],
          [
            60,
            60,
          ],
        ],
        "type": "LineString",
      },
      "hidden": false,
      "hoveredByMe": false,
      "id": "id1",
      "idx": "",
      "name": null,
      "parent": {
        "children": [],
        "id": "",
        "parent": null,
      },
      "selectedByMe": false,
      "selectedByPeers": null,
      "type": "feature",
    },
    "geometry": {
      "coordinates": [
        60,
        60,
      ],
      "type": "Point",
    },
    "handleType": "vertex",
    "id": "id1-handle1",
    "type": "handle",
  },
]
`;

exports[`single point snapshot 1`] = `
[
  {
    "active": false,
    "children": [],
    "color": "#000000",
    "geometry": {
      "coordinates": [
        50,
        50,
      ],
      "type": "Point",
    },
    "hidden": false,
    "hoveredByMe": false,
    "id": "id1",
    "idx": "",
    "name": null,
    "parent": {
      "children": [],
      "id": "",
      "parent": null,
    },
    "selectedByMe": false,
    "selectedByPeers": null,
    "type": "feature",
  },
]
`;
